host/sgx/enter.c

```diff
@@ -128,34 +119,37 @@ int __oe_host_stack_bridge(
 
     return ret;
 }
-
-/**
- * Size of ocall buffers passed in ecall_contexts. Large enough for most ocalls.
- * If an ocall requires more than this size, then the enclave will make an
- * ocall to allocate the buffer instead of using the ecall_context's buffer.
- * Note: Currently, quotes are about 10KB.
- */
-#define OE_DEFAULT_OCALL_BUFFER_SIZE (16 * 1024)
-
 /**
  * Setup the ecall_context.
+ * This function should never be inline so that it can record the
+ * caller's stack frame. The stack frame information is used to stitch
+ * the stack upon an enclave entry.
  */
-OE_INLINE void _setup_ecall_context(oe_ecall_context_t* ecall_context)
+OE_NEVER_INLINE
+void oe_setup_ecall_context(oe_ecall_context_t* ecall_context)
 {
     oe_thread_binding_t* binding = oe_get_thread_binding();
+    uint64_t* caller_frame = __builtin_frame_address(1);
+
     if (binding->ocall_buffer == NULL)
     {
         // Lazily allocate buffer for making ocalls. Bound to the tcs.
         // Will be cleaned up by enclave during termination.
         binding->ocall_buffer = malloc(OE_DEFAULT_OCALL_BUFFER_SIZE);
         binding->ocall_buffer_size = OE_DEFAULT_OCALL_BUFFER_SIZE;
     }
+
     ecall_context->ocall_buffer = binding->ocall_buffer;
     ecall_context->ocall_buffer_size = binding->ocall_buffer_size;
+
+    /* Record caller's stack frame */
+    ecall_context->debug_eenter_rbp = caller_frame[0];
```

Comment:
To retain current behavior (in both Linux and Windows), it is best set these fields only in case of vdso.